drivers/flash/flash_stm32h7x.c

```diff
@@ -0,0 +1,534 @@
+	for (i = 0; i < n; ++i) {
+		flash[i] = data[i];
+		wait_write_queue(dev, offset);
+	}
+
+	/* Wait until the BSY bit is cleared */
+	rc = flash_stm32h7_wait_flash_idle(dev);
+
+	/* Clear the PG bit */
+	*cr &= (~FLASH_CR_PG);
+
+	return rc;
+}
+
+int flash_stm32_write_range(const struct device *dev, unsigned int offset,
+			    const void *data, unsigned int len)
+{
+	int rc = 0;
+	int i = 0;
+	int j;
+	const uint8_t ndwords = FLASH_NB_32BITWORD_IN_FLASHWORD / 2;
+	const uint8_t nbytes = FLASH_NB_32BITWORD_IN_FLASHWORD * 4;
+	uint8_t unaligned_datas[nbytes];
+	int suboffset = (offset % nbytes);
+
+	if (suboffset != 0) {
+		/* Offset is not aligned on FLASHWORD boundary */
+		offset -= suboffset;
+		memset(unaligned_datas, 0xff, sizeof(unaligned_datas));
+		for (i = 0; i < len && i < nbytes; ++i) {
+			unaligned_datas[suboffset + i] = ((uint8_t *)data)[i];
+		}
+		rc = write_ndwords(dev, offset,
+				   (const uint64_t *)unaligned_datas,
+				   ndwords);
+		i = 32;
+		offset += 32;
+	}
+
+	for (; i < len && i + 32 <= len; i += 32, offset += 32U) {
+		rc = write_ndwords(dev, offset,
+				   (const uint64_t *) data + (i >> 3),
+				   ndwords);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	/* Handle the remaining bytes if length is not aligned on
+	 * FLASH_NB_32BITWORD_IN_FLASHWORD
+	 */
+	if (i < len) {
+		memset(unaligned_datas, 0xff, sizeof(unaligned_datas));
+		for (j = 0; j < len - i; ++j) {
+			unaligned_datas[j] = ((uint8_t *)data)[i + j];
+		}
+		rc = write_ndwords(dev, offset,
+				   (const uint64_t *)unaligned_datas,
+				   ndwords);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+static int flash_stm32h7_write_protection(const struct device *dev, bool enable)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+
+	int rc = 0;
+
+	flash_stm32_sem_take(dev);
+
+	if (enable) {
+		rc = flash_stm32h7_wait_flash_idle(dev);
+		if (rc) {
+			flash_stm32_sem_give(dev);
+			return rc;
+		}
+	}
+
+	/* Bank 1 */
+	if (enable) {
+		regs->CR1 |= FLASH_CR_LOCK;
+	} else {
+		if (regs->CR1 & FLASH_CR_LOCK) {
+			regs->KEYR1 = FLASH_KEY1;
+			regs->KEYR1 = FLASH_KEY2;
+		}
+	}
+	/* Bank 2 */
+	if (enable) {
+		regs->CR2 |= FLASH_CR_LOCK;
+	} else {
+		if (regs->CR2 & FLASH_CR_LOCK) {
+			regs->KEYR2 = FLASH_KEY1;
+			regs->KEYR2 = FLASH_KEY2;
+		}
+	}
+
+	if (enable) {
+		LOG_DBG("Enable write protection");
+	} else {
+		LOG_DBG("Disable write protection");
+	}
+
+	flash_stm32_sem_give(dev);
+
+	return rc;
+}
+
+static void flash_stm32h7_flush_caches(const struct device *dev,
+				       off_t offset, size_t len)
+{
+	ARG_UNUSED(dev);
+	SCB_InvalidateDCache_by_Addr((uint32_t *)(CONFIG_FLASH_BASE_ADDRESS
+						  + offset), len);
+}
+
+
+static int flash_stm32h7_erase(const struct device *dev, off_t offset,
+			       size_t len)
+{
+	int rc;
+	off_t flush_offset = get_sector(offset) * FLASH_SECTOR_SIZE;
+	size_t flush_len = (((get_sector(offset + len) + 1) * FLASH_SECTOR_SIZE)
```

Comment:
I'm also confused here. Why length to be used in SCB_InvalidateDCache_by_Addr would be different from len?

Reply:
Same issue as above.
Although I suspect there is still a bug lying somewhere because if I do a write_pattern with a length of 256 at offset 0, then a write_pattern of length 32 at offset 0, I still see the data at offet 32 to 63.

Reply:
Ok, I saw that as well.
I would tend to say this is related to a bug in upper layers since this cannot be seen if `flash erase` is run manually between two `flash write_pattern`